pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,326 @@
+func (b *Builder) WithServerTransportCredentials(insecureSkipVerify bool, certPool *x509.CertPool) {
+	if b.tlsConfig == nil {
+		b.tlsConfig = &tls.Config{}
+	}
+	if insecureSkipVerify {
+		b.tlsConfig.InsecureSkipVerify = true
+	}
+	if certPool != nil {
+		b.tlsConfig.RootCAs = certPool
+	}
+}
+
+// GetServerTransportCredentials ...
+func (b *Builder) GetServerTransportCredentials() (bool, *x509.CertPool) {
+	return b.tlsConfig.InsecureSkipVerify, b.tlsConfig.RootCAs
+}
+
+// WithClientTransportCredentials builds transport credentials for a gRPC client uses to connect to service
+func (b *Builder) WithClientTransportCredentials(certs ...tls.Certificate) {
+	b.tlsConfig.Certificates = certs
+}
+
+// GetClientTransportCredentials ...
+func (b *Builder) GetClientTransportCredentials() []tls.Certificate {
+	return b.tlsConfig.Certificates
+}
+
+// WithClientCredentials builds transport credentials for a gRPC client
+func (b *Builder) WithClientCredentials(c credentials.PerRPCCredentials) {
+	b.credentials = c
+}
+
+// GetClientCredentials ...
+func (b *Builder) GetClientCredentials() credentials.PerRPCCredentials {
+	return b.credentials
+}
+
+// SetConnInfo allows passing in the dns and port for the connection, providing
+// flexibility if a consumer wants to set a default and or override
+func (b *Builder) SetConnInfo(dns, port string, tls bool) error {
+	u, err := url.Parse(dns)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid dns: %s", dns))
+	}
+	if u.Scheme != "" {
+		// This will cause the connection to fail silently with a cryptic "context.Deadline exceeded" so we validate for it here.
+		return errors.New(fmt.Sprintf("grpc connection dns must not contain a scheme/protocol: %s", dns))
+	}
+	b.dns = &dns
+
+	i, err := strconv.ParseUint(port, 10, 16)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid port number: %s", port))
+	}
+	p := uint16(i)
+	b.port = &p
+
+	return nil
+}
+
+// GetConnInfo returns the dns and port that were set, and errors if either
+// were null, allowing verfication prior to attempting the connection
+func (b *Builder) GetConnInfo() (dns string, port uint16, err error) {
+	if b.dns == nil || b.port == nil {
+		return "", 0, errors.New("Connection info not fully set")
+	}
+	return *b.dns, *b.port, nil
+}
+
+// SetConnAddr sets connection information as well as configures some common tls and authentication options
+// based on a provided connection string.
+// the format is tls://hostname:port (or tcp://hostname:port if developing locally without TLS)
+// Optional TLS parameters:
+//   skip_verify=true        ignore server CA verification.
+//   ca_file=./filename.pem  CA of service for verification.
+func (b *Builder) SetConnAddr(addr string) error {
+	cfg, err := ReadConfig(addr)
+	if err != nil {
+		return err
+	}
+	cfg.ApplyToBuilder(b)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Builder) WithFS(fs fs.FS) {
```

Comment:
I like this because suddenly it opens up a world of sharing our variables via an S3 or SSM FS, and not just via env vars.